radicle-httpd/src/api/json.rs

```diff
@@ -64,6 +71,53 @@ pub(crate) fn blob<T: AsRef<[u8]>>(blob: &Blob<T>, path: &str) -> Value {
     response
 }
 
+/// Creates a [`FileStream`] from `blob`, using `tmp_path` as the file.
+#[allow(dead_code)]
+pub(crate) fn blob_stream<P: AsRef<Path>>(
+    blob: &[u8],
+    tmp_path: P,
+    chunk_sz: usize,
+) -> Result<FileStream, crate::Error> {
+    let tmp_path = tmp_path.as_ref();
+    let mut write_f = File::create(tmp_path)?;
+    write_f.write_all(blob)?;
+    FileStream::new(tmp_path, chunk_sz)
+}
+
+/// Represents a read stream of a file.
+pub(crate) struct FileStream {
+    f: File,
+    chunk_sz: usize,
+}
```

Comment:
I would have expected that you could have made a stream of the byte slice. Why are we opting to use a `File` here?

I'm pretty sure it's something along the lines of:
```rust
use futures::stream;

pub fn blob_stream(bs: &[u8]) -> impl Stream {
    stream::iter(bs)
}
```

The caller of `blob` can decide on their own chunking by calling [`chunks`](https://docs.rs/futures-util/0.3.26/futures_util/stream/trait.StreamExt.html#method.chunks).

Reply:
The essence of this patch is to support streaming from a file. The [motivation](https://github.com/radicle-dev/heartwood/pull/331#discussion_r1110054271) was to store a big blobs into a tmp file so that we can reduce memory consumption. Then `httpd` would stream the blob from the tmp file.

Reply:
I see. It _seems_ a bit weird that you have to have the bytes in memory to avoid them being in memory -- but perhaps it makes sense.

Regardless, I don't think we need to roll our own here. I believe you could use a combination of [`futures::io::BufReader`](https://docs.rs/futures/0.3.26/futures/io/struct.BufReader.html), [`futures::io::AsyncBufReadExt::lines`](https://docs.rs/futures/0.3.26/futures/io/trait.AsyncBufReadExt.html#method.lines), and [`futures::StreamExt::chunks`](https://docs.rs/futures-util/0.3.26/futures_util/stream/trait.StreamExt.html#method.chunks) to achieve this. We would create the file and pass it to `BufReader::new`.

Also, because you're using a temporary directory, be _very_ careful that you hold onto the `TempDir` value until the stream is finished, otherwise it will be dropped and the directory will be removed.

Reply:
> Also, because you're using a temporary directory, be _very_ careful that you hold onto the `TempDir` value until the stream is finished, otherwise it will be dropped and the directory will be removed.

yeah, I think this line holds a reference to the underlying `TempDir` in the test: 

```
let tmp_file = tmp_dir.path().join("test_blob_stream");
```